### ISSUE TITLE
etcd-launcher does not successfully recover from PVC deletion

### ISSUE THREAD

**Issue by embik** [kubermatic/kubermatic#9020]
### What happened?

This was originally raised as a PR (#9019).

When deleting a single PVC for etcd when `etcd-launcher` is enabled, the fresh replacement pod does not successfully start but instead fails.

```
022-02-16 12:53:27.829464 I | etcdserver: starting server... [version: 3.4.3, cluster version: to_be_decided]
2022-02-16 12:53:27.832051 I | embed: ClientTLS: cert = /etc/etcd/pki/tls/etcd-tls.crt, key = /etc/etcd/pki/tls/etcd-tls.key, trusted-ca = /etc/etcd/pki/ca/ca.crt, client-cert-auth = true, crl-file =
2022-02-16 12:53:27.832081 I | embed: listening for peers on 10.44.5.194:2381
2022-02-16 12:53:27.832289 I | embed: listening for metrics on http://127.0.0.1:2378
2022-02-16 12:53:27.832428 I | embed: listening for metrics on http://10.44.5.194:2378
2022-02-16 12:53:27.859967 I | rafthttp: started HTTP pipelining with peer f06ca2eeaf5c8ad8
raft2022/02/16 12:53:27 INFO: a4adad3686f23568 [term: 0] received a MsgHeartbeat message with higher term from f06ca2eeaf5c8ad8 [term: 8]
raft2022/02/16 12:53:27 INFO: a4adad3686f23568 became follower at term 8
raft2022/02/16 12:53:27 tocommit(2292) is out of range [lastIndex(0)]. Was the raft log corrupted, truncated, or lost?
panic: tocommit(2292) is out of range [lastIndex(0)]. Was the raft log corrupted, truncated, or lost?

goroutine 127 [running]:
log.(*Logger).Panicf(0xc0000c4aa0, 0x10c8397, 0x5d, 0xc000217c40, 0x2, 0x2)
	/usr/local/go/src/log/log.go:219 +0xc1
go.etcd.io/etcd/raft.(*DefaultLogger).Panicf(0x1aa0040, 0x10c8397, 0x5d, 0xc000217c40, 0x2, 0x2)
	/tmp/etcd-release-3.4.3/etcd/release/etcd/raft/logger.go:127 +0x60
go.etcd.io/etcd/raft.(*raftLog).commitTo(0xc000219c00, 0x8f4)
	/tmp/etcd-release-3.4.3/etcd/release/etcd/raft/log.go:203 +0x131
go.etcd.io/etcd/raft.(*raft).handleHeartbeat(0xc000183680, 0x8, 0xa4adad3686f23568, 0xf06ca2eeaf5c8ad8, 0x8, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/tmp/etcd-release-3.4.3/etcd/release/etcd/raft/raft.go:1396 +0x54
go.etcd.io/etcd/raft.stepFollower(0xc000183680, 0x8, 0xa4adad3686f23568, 0xf06ca2eeaf5c8ad8, 0x8, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/tmp/etcd-release-3.4.3/etcd/release/etcd/raft/raft.go:1341 +0x480
go.etcd.io/etcd/raft.(*raft).Step(0xc000183680, 0x8, 0xa4adad3686f23568, 0xf06ca2eeaf5c8ad8, 0x8, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/tmp/etcd-release-3.4.3/etcd/release/etcd/raft/raft.go:984 +0x1113
go.etcd.io/etcd/raft.(*node).run(0xc0002d8420)
	/tmp/etcd-release-3.4.3/etcd/release/etcd/raft/node.go:352 +0xab6
created by go.etcd.io/etcd/raft.RestartNode
	/tmp/etcd-release-3.4.3/etcd/release/etcd/raft/node.go:240 +0x33c
{"level":"info","time":"2022-02-16T12:53:28.840Z","caller":"etcd-launcher/main.go:139","msg":"[https://etcd-1.etcd.cluster-g4mp4z6w52.svc.cluster.local:2381] is a member"}
{"level":"panic","time":"2022-02-16T12:53:28.840Z","caller":"etcd-launcher/main.go:174","msg":"exit status 2"}
panic: exit status 2

goroutine 1 [running]:
go.uber.org/zap/zapcore.(*CheckedEntry).Write(0xc000036780, {0x0, 0x0, 0x0})
	go.uber.org/zap@v1.19.1/zapcore/entry.go:232 +0x446
go.uber.org/zap.(*SugaredLogger).log(0xc0000b77e8, 0x4, {0x0, 0xe}, {0xc00098faf0, 0x1, 0x1}, {0x0, 0x0, 0x0})
	go.uber.org/zap@v1.19.1/sugar.go:227 +0xee
go.uber.org/zap.(*SugaredLogger).Panic(...)
	go.uber.org/zap@v1.19.1/sugar.go:123
main.main()
	k8c.io/kubermatic/v2/cmd/etcd-launcher/main.go:174 +0x1396
```

### What should happen?

`etcd-launcher` is successfully adding the fresh pod as new node to recover from data loss.

### How to reproduce?

1. Create cluster.
2. Enable feature gate `etcdLauncher` on cluster.
3. Force deletion of a PV attached to an etcd PVC (delete the PVC and delete matching Pod).
4. See that Pod is failing.

### Environment

I tested this on the following environment. @furkhat probably used a different version.

- KKP Version: `master`
- Shared or separate master/seed clusters?: shared
- Master/seed cluster provider: -

**Comment by embik** [comment]
Actually, it does look like it's successful? Not sure, needs some more investigation. When I hit the problem, I deleted the PVC, not the PV. The docs specifically say that it reacts to the PV being unavailable.

I'll check again with the PVC deletion and also check if we can extend the reaction of the PV loss to the PVC, if that is needed.

**Comment by squ94wk** [comment]
I believe there's recovery code only for PVs with Phase "Lost".
If you delete the PVC, the PV gets unbound and reclaimed (probably deleted), so the code wouldn't trigger.

What the logs are saying is that the rest of the cluster knows this member and expect its log to be in a "recent" state, not 0. So the cluster doesn't allow the member to join. Once you remove the member it joins again.

**Comment by embik** [comment]
@squ94wk that is correct (except for the "Lost" phase being on PVCs, not PVs).

Curiously, this does not happen with etcd 3.5, only with 3.4. With the known problems of etcd 3.5, I'm not sure yet if they fixed this problem in 3.5 or it's just even worse because etcd starts up and serves incorrect data. Investigating that.
